File: code/Java/src/main/java/Model/Position.java
package Model;

import java.util.Objects;

public class Position {
    private final Coordinate coordinate;
    private final Direction direction;

    public Position(int x, int y, String directionText) {
        this.coordinate = new Coordinate(x, y);
        this.direction = switch (directionText) {
            case "N" -> Direction.NORTH;
            case "E" ->Direction.EAST;
            case "S" ->Direction.SOUTH;
            case "W" ->Direction.WEST;
            default -> Direction.NORTH;
        };
    }

    public Position turnLeft() {
        return new Position(this.coordinate.x(), this.coordinate.y(), this.direction.turnLeft().toString());
    }

    public Position turnRight() {
        return new Position(this.coordinate.x(), this.coordinate.y(), this.direction.turnRight().toString());
    }

    public Position moveForward() {
        var coordinate = switch (this.direction) {
            case NORTH -> this.coordinate.moveNorth();
            case EAST -> this.coordinate.moveEast();
            case SOUTH -> this.coordinate.moveSouth();
            case WEST -> this.coordinate.moveWest();
        };
        return new Position(coordinate.x(), coordinate.y(), this.direction.toString());
    }

    @Override
    public String toString() {
        return String.format("%s %s", coordinate.toString(), direction.toString());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Position position = (Position) o;
        return Objects.equals(coordinate, position.coordinate) &&
                direction == position.direction;
    }

    @Override
    public int hashCode() {
        return Objects.hash(coordinate, direction);
    }
}

File: code/CSharp/ConnascentMarsRover/Source/Model/Position.cs
using System;
using static Source.Model.Direction;

namespace Source.Model
{
    public class Position
    {
        private readonly Coordinate _coordinate;
        private readonly Direction _direction;

        public Position(Coordinate coordinate, Direction direction)
        {
            _coordinate = coordinate;
            _direction = direction;
        }

        public Position MoveForward()
        {
            return _direction switch
            {
                NORTH => new Position(_coordinate.MoveNorth(), _direction),
                EAST => new Position(_coordinate.MoveEast(), _direction),
                SOUTH => new Position(_coordinate.MoveSouth(), _direction),
                WEST => new Position(_coordinate.MoveWest(), _direction),
                _ => throw new ArgumentOutOfRangeException()
            };
        }

        public Position TurnLeft()
        {
            return new Position(_coordinate, _direction.TurnLeft());
        }

        public Position TurnRight()
        {
            return new Position(_coordinate, _direction.TurnRight());
        }

        protected bool Equals(Position other)
        {
            return Equals(_coordinate, other._coordinate) && _direction == other._direction;
        }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != this.GetType()) return false;
            return Equals((Position) obj);
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(_coordinate, (int) _direction);
        }

        public override string ToString()
        {
            var direction = _direction.ToString().Substring(0, 1);
            return $"{_coordinate} {direction}";
        }
    }
}